File: 295. Find Median from Data Stream/solution.py
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.minHeap = []
        self.maxHeap = []

    def addNum(self, num: int) -> None:
        # Time complexity - O(logN)
        # Space complexity - O(N)
        heapq.heappush(self.minHeap, num)
        if len(self.minHeap) > len(self.maxHeap):
            num = heapq.heappop(self.minHeap)
            heapq.heappush(self.maxHeap, -num)
            
        if self.minHeap and self.maxHeap and self.minHeap[0] < -self.maxHeap[0]:
            minNum = heapq.heappop(self.minHeap)
            maxNum = heapq.heappushpop(self.maxHeap, -minNum)
            heapq.heappush(self.minHeap, -maxNum)
            
    def findMedian(self) -> float:
        # Time complexity - O(1)
        # Space complexity - O(1)
        if len(self.minHeap) == len(self.maxHeap):
            return (self.minHeap[0] - self.maxHeap[0])/2.0
        else:
            return -1.0 * self.maxHeap[0]
        
# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 295. Find Median from Data Stream/solution.java
class MedianFinder {
    private PriorityQueue<Integer> lower;
    private PriorityQueue<Integer> upper;
    
    /** initialize your data structure here. */
    public MedianFinder() {
        lower = new PriorityQueue<Integer>((a, b) -> b - a);
        upper = new PriorityQueue<Integer>((a, b) -> a - b);
    }
    
    public void addNum(int num) {
        // insert the number in upper half first
        upper.add(num);
        
        // rebalance the size of lower == upper or lower - 1 == upper
        num = upper.remove();
        lower.add(num);
        
        if (lower.size() - 1 > upper.size()) {
            num = lower.remove();
            upper.add(num);
        }
    }
    
    public double findMedian() {
        if (lower.size() == 0) {
            return 0.0;
        }
        double sol = lower.peek();
        
        if (lower.size() == upper.size()) {
            sol += upper.peek();
            sol /= 2;
        }
        
        return sol;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
